### PR TITLE
Migration of ChannelStore Part 1

### LINKED ISSUE #15129: Migrate AppError to error from store/sqlstore/channel_store.go part 1
Mattermost is migrating all usages of the `AppError` type to the more idiomatic `error` type. `AppError` is a loaded error type containing fields like HTTP Status code which is not really necessary for errors coming from the DB layer. We should migrate to using normal types satisfying the error interface.

This ticket is about migrating all usages of AppError from `store/sqlstore/channel_store.go part 1` to return a normal error. The migration should use the error types defined in `store/errors.go`. If the need should arise to add another specific error type, feel free to add it. For any generic error that is not something like an invalid input, or something not found, we can just wrap the error using `errors.Wrap(f)` and return it.

And then in the app layer, we would need to convert the `error` type to the usual `AppError`. Essentially moving the `AppError` creation from the store layer to the app layer. This means we need a switch case to check for the type of error and appropriately convert to the correct AppError. It can be something like:

```go
if _, err := a.Srv().Store.Bot().Update(bot); err != nil {
	var nfErr *store.ErrNotFound
	var appErr *model.AppError
	switch {
	case errors.As(err, &nfErr):
		return model.MakeBotNotFoundError(nfErr.Id)
	case errors.As(err, &appErr): // in case we haven't converted to plain error.
		return appErr
	default: // last fallback in case it doesn't map to an existing app error.
		return model.NewAppError("DeleteBotIconImage", "app.bot.patchbot.internal_error", nil, err.Error(), http.StatusInternalServerError)
	}
}
```

Also note that the error codes need to be changed from being a store level code (`store.<>.edit.error`) to something like `app.<>.edit.internal_error`.

Example Pull Request: #14339

If you have questions about the ticket or you need any help, feel free to contact `agniva.de.sarker` in https://community.mattermost.com/

----------------------------------

If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

JIRA: https://mattermost.atlassian.net/browse/MM-27322

### PR DESCRIPTION
**Pull request by rvillablanca**
#### Summary

First set of methods in `ChannelStore` that return error instead of `*model.AppError`

#### Ticket Link

  Fixes #15129

### CODE REVIEW

**Comment by rvillablanca** [comment]
@agnivade First part of ChannelStore :tada:

File: app/channel.go [hunk 1]
```diff
@@ -2396,8 +2428,8 @@ func (a *App) MoveChannel(team *model.Team, channel *model.Channel, user *model.
 		return err
 	}
 
-	if appErr := a.Srv().Store.Channel().UpdateSidebarChannelCategoryOnMove(channel, team.Id); appErr != nil {
-		return appErr
+	if nErr := a.Srv().Store.Channel().UpdateSidebarChannelCategoryOnMove(channel, team.Id); nErr != nil {
+		return model.NewAppError("MoveChannel", "app.channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
```
Comment (agnivade) [inline app/channel.go:2432]: This should be `nErr.Error()`. Also see: https://github.com/mattermost/mattermost-server/pull/15268. This is a common bug that happens during copy-paste. Just a small note for your future PRs :)

File: app/team.go [hunk 2]
```diff
@@ -1381,8 +1381,9 @@ func (a *App) PermanentDeleteTeam(team *model.Team) *model.AppError {
 	}
 
 	if channels, err := a.Srv().Store.Channel().GetTeamChannels(team.Id); err != nil {
-		if err.Id != "app.channel.get_channels.not_found.app_error" {
-			return err
+		var nfErr *store.ErrNotFound
+		if !errors.As(err, &nfErr) {
+			return model.NewAppError("PermanentDeleteTeam", "app.channel.get_channels.get.app_error", nil, err.Error(), http.StatusInternalServerError)
```
Comment (agnivade) [inline app/team.go:1386]: `nfErr.Error()`

File: store/sqlstore/channel_store.go [hunk 3]
```diff
@@ -1387,12 +1387,12 @@ type channelIdWithCountAndUpdateAt struct {
...
 }
 
-func (s SqlChannelStore) GetChannelCounts(teamId string, userId string) (*model.ChannelCounts, *model.AppError) {
+func (s SqlChannelStore) GetChannelCounts(teamId string, userId string) (*model.ChannelCounts, error) {
 	var data []channelIdWithCountAndUpdateAt
 	_, err := s.GetReplica().Select(&data, "SELECT Id, TotalMsgCount, UpdateAt FROM Channels WHERE Id IN (SELECT ChannelId FROM ChannelMembers WHERE UserId = :UserId) AND (TeamId = :TeamId OR TeamId = '') AND DeleteAt = 0 ORDER BY DisplayName", map[string]interface{}{"TeamId": teamId, "UserId": userId})
 
 	if err != nil {
-		return nil, model.NewAppError("SqlChannelStore.GetChannelCounts", "store.sql_channel.get_channel_counts.get.app_error", nil, "teamId="+teamId+", userId="+userId+", err="+err.Error(), http.StatusInternalServerError)
+		return nil, errors.Wrapf(err, "failed to get channels counts with teamId=%s and userId=%s", teamId, userId)
```
Comment (agnivade) [inline store/sqlstore/channel_store.go:1395]: Nit: "channels count" is more correct. Actually, even `model.ChannelCounts` is wrong. But we can't break semver now :)

File: store/sqlstore/channel_store.go [hunk 4]
```diff
@@ -2508,13 +2506,12 @@ func (s SqlChannelStore) GetChannelsByIds(channelIds []string, includeDeleted bo
 	_, err := s.GetReplica().Select(&channels, query, params)
 
 	if err != nil {
-		mlog.Error("Query error getting channels by ids", mlog.Err(err))
```
Comment (agnivade) [inline store/sqlstore/channel_store.go:2511]: Oh dear, we were logging and returning an error. Great work cleaning this up.

File: app/channel.go (hunk 1 above)
Comment (rvillablanca) [reply app/channel.go:2432]: Yep my bad :facepalm: sorry about that panic. I will keep my eyes on these cases

File: app/team.go (hunk 2 above)
Comment (rvillablanca) [reply app/team.go:1386]: This case is ok, notice the `!` before `errors.As(...`

**Comment by rvillablanca** [comment]
I'm done with all suggestions @agnivade. This time I decided to keep conversations open bacause I believe they should be closed by whoever open them when they are considered ok.
Thanks for the review!

File: app/team.go (hunk 2 above)
Comment (agnivade) [reply app/team.go:1397]: Ah my bad.

**Review by agnivade — APPROVED** [review]
Looks great!

**Review by jespino — APPROVED** [review]
Looks good to me. Thanks @rvillablanca! :confetti_ball:

**Comment by rvillablanca** [comment]
Seems there was some code splitting so I fixed the merge conflicts

**Comment by agnivade** [comment]
@rvillablanca - this is the error

```
/go/src/github.com/mattermost/mattermost-server/bin/mmgotool i18n extract --portal-dir=""
7289a7290,7293
>     "id": "store.sql_channel.sidebar_categories.app_error",
>     "translation": ""
>   },
>   {
```

I haven't dug deep, but it doesn't look like this needs EE repo access.

**Comment by rvillablanca** [comment]
Ok let me check

**Comment by rvillablanca** [comment]
@agnivade Fixed

**Comment by agnivade** [comment]
@rvillablanca - You are a core committer now. So you are most welcome to merge your own PRs :D And as a bonus, the EETest label need not be manually added every time now.

**Comment by rvillablanca** [comment]
@agnivade That's great :tada: unfortunately I still don't have write permission to this repository

**Pull request merged**